### PR TITLE
Improve exception logging

### LINKED ISSUE #5677: CodeIgniter\Debug\Exceptions::exceptionHandler() - Add File and Line to message
### PHP Version

8.0

### CodeIgniter4 Version

4.1.8

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

Windows

### Which server did you use?

apache

### Database

_No response_

### What happened?

Sort of a bug, sort of a feature request.

In some cases Exception trace doesn't include any useful information. 

In CodeIgniter\Debug\Exceptions::exceptionHandler() the log_message() function is called as follows:

```
            log_message('critical', $exception->getMessage() . "\n{trace}", [
                'trace' => $exception->getTraceAsString(),
            ]);
```

It would be helpful to include the file and line in the message.

### Steps to Reproduce

Create a controller method and put the following lines:

set_time_limit(2);
sleep(5);

### Expected Output

The log file produces:

```
CRITICAL - 2022-02-10 14:28:06 --> Maximum execution time of 2 seconds exceeded
#0 [internal function]: CodeIgniter\Debug\Exceptions->shutdownHandler()
#1 {main}
```

This isn't helpful to know where the timeout is occurring.

### Anything else?

By changing the CodeIgniter\Debug\Exceptions::exceptionHandler()  log_message() function to include file and line:

```
            log_message('critical', $exception->getMessage().' called in '.$this->cleanPath($exception->getFile()).' line '.$exception->getLine() . "\n{trace}", [
                'trace' => $exception->getTraceAsString(),
            ]);
```

We now get useful information to where the problem is.

```
CRITICAL - 2022-02-10 14:29:05 --> Maximum execution time of 2 seconds exceeded called in APPPATH\Controllers\Admin\Test.php line 14
#0 [internal function]: CodeIgniter\Debug\Exceptions->shutdownHandler()
#1 {main}
```

Perhaps there is a better way to do this.. maybe even a configurable option to turn this on?

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
Closes #5677 

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Debug/Exceptions.php [hunk 1]
```diff
@@ -434,4 +437,52 @@ public static function highlightFile(string $file, int $lineNumber, int $lines =
...
+
+                    case is_array($value):
+                        return $value !== [] ? '[...]' : '[]';
+
+                    case $value === null:
+                        return 'null';
+
+                    case is_resource($value):
+                        return sprintf('resource (%s)', get_resource_type($value));
+
```
Comment (sclubricants) [inline system/Debug/Exceptions.php:469]:
```php
					case is_string($value):
						return "'".clean_path($value)."'";
```

**Comment by kenjis** [comment]
Why this PR is refactoring?
It seems it is enhancement.

File: tests/system/Debug/ExceptionsTest.php [hunk 2]
```diff
@@ -61,4 +61,20 @@ public function testDetermineCodes(): void
...
+        $renderedBacktrace = $renderer($exception->getTrace());
+        $renderedBacktrace = explode("\n", $renderedBacktrace);
+
+        foreach ($renderedBacktrace as $trace) {
+            $this->assertMatchesRegularExpression(
+                '/^\s*\d* .+(?:\(\d+\))?: \S+(?:(?:\->|::)\S+)?\(.*\)$/',
+                $trace
+            );
+        }
+    }
```
Comment (kenjis) [inline tests/system/Debug/ExceptionsTest.php:79]: I don't understand what's tested.
I think we need tests to prove the changed output.

**Comment by paulbalandan** [comment]
> Why this PR is refactoring? It seems it is enhancement.

Changed title.

File: tests/system/Debug/ExceptionsTest.php (hunk 2 above)
Comment (paulbalandan) [reply tests/system/Debug/ExceptionsTest.php:79]: That's true. But I added a new method. At least I don;t want it to be uncovered by tests.

**Comment by kenjis** [comment]
Before:
```
CRITICAL - 2022-02-15 02:10:32 --> Cannot modify header information - headers already sent
#0 [internal function]: CodeIgniter\Debug\Exceptions->errorHandler(2, 'Cannot modify h...', '/Users/kenji/wo...', 11)
#1 /Users/kenji/work/codeigniter/CodeIgniter4/app/Controllers/Home.php(11): setcookie('printer', '37')
#2 /Users/kenji/work/codeigniter/CodeIgniter4/system/CodeIgniter.php(834): App\Controllers\Home->index()
#3 /Users/kenji/work/codeigniter/CodeIgniter4/system/CodeIgniter.php(421): CodeIgniter\CodeIgniter->runController(Object(App\Controllers\Home))
#4 /Users/kenji/work/codeigniter/CodeIgniter4/system/CodeIgniter.php(329): CodeIgniter\CodeIgniter->handleRequest(NULL, Object(Config\Cache), false)
#5 /Users/kenji/work/codeigniter/CodeIgniter4/public/index.php(37): CodeIgniter\CodeIgniter->run()
#6 /Users/kenji/work/codeigniter/CodeIgniter4/system/Commands/Server/rewrite.php(43): require_once('/Users/kenji/wo...')
#7 {main}
```

After:
```
CRITICAL - 2022-02-15 02:12:17 --> Cannot modify header information - headers already sent
in APPPATH/Controllers/Home.php on line 11.
 1 [internal function]: CodeIgniter\Debug\Exceptions->errorHandler(2, 'Cannot modify header information - headers already sent', 'APPPATH/Controllers/Home.php', 11)
 2 APPPATH/Controllers/Home.php(11): setcookie('printer', '37')
 3 SYSTEMPATH/CodeIgniter.php(834): App\Controllers\Home->index()
 4 SYSTEMPATH/CodeIgniter.php(421): CodeIgniter\CodeIgniter->runController(Object(App\Controllers\Home))
 5 SYSTEMPATH/CodeIgniter.php(329): CodeIgniter\CodeIgniter->handleRequest(null, Object(Config\Cache), false)
 6 FCPATH/index.php(37): CodeIgniter\CodeIgniter->run()
 7 SYSTEMPATH/Commands/Server/rewrite.php(43): require_once('FCPATH/index.php')
```

File: system/Debug/Exceptions.php [hunk 3]
```diff
@@ -434,4 +437,55 @@ public static function highlightFile(string $file, int $lineNumber, int $lines =
...
+        foreach ($backtrace as $index => $trace) {
+            $frame = $trace + ['file' => '[internal function]', 'line' => '', 'class' => '', 'type' => '', 'args' => []];
+
+            if ($frame['file'] !== '[internal function]') {
+                $frame['file'] = sprintf('%s(%s)', $frame['file'], $frame['line']);
+            }
+
+            unset($frame['line']);
+            $idx = $index;
+            $idx = str_pad((string) ++$idx, 2, ' ', STR_PAD_LEFT);
```
Comment (kenjis) [inline system/Debug/Exceptions.php:454]: Why do you start with 1, not 0?

File: tests/system/Debug/ExceptionsTest.php (hunk 2 above)
Comment (kenjis) [reply tests/system/Debug/ExceptionsTest.php:79]: Okay, it is difficult to write tests for this.

**Comment by kenjis** [comment]
@paulbalandan The log format will be changed. If a user watches log files and parse it to check the error, it may break.
I think it is better to add to changelog.

File: system/Debug/Exceptions.php (hunk 3 above)
Comment (paulbalandan) [reply system/Debug/Exceptions.php:454]: I feel that's more "natural" way of counting. No offense to zero-based array counting. 😂

Comment (sclubricants) [reply system/Debug/Exceptions.php:454]:
```php
$idx = str_pad((string) ++$idx, 2, '#', STR_PAD_LEFT);
```
This is like the original.

**Review by kenjis — APPROVED** [review]

File: system/Debug/Exceptions.php (hunk 3 above)
Comment (paulbalandan) [reply system/Debug/Exceptions.php:454]: > This is like the original.

This would fail on idx >= 10. https://3v4l.org/DAqmo

Comment (sclubricants) [reply system/Debug/Exceptions.php:454]: https://3v4l.org/dFGjg

echo '#'.str_pad(++$idx, 3, ' ', STR_PAD_RIGHT).'Some more text', PHP_EOL;

**Pull request merged**